### ISSUE TITLE
Rule Proposal: Check for global references inside `render`

### ISSUE THREAD

**Issue by mridgway** [yannickcr/eslint-plugin-react#246]
When using React for server rendering, it can be problematic when a component relies on `window` or `document` or other globals inside of render. It will cause checksum issues when you get to the client. Instead of checking whether you are on the client within `render`, it's better to use the `componentDidMount` lifecycle event to set a state which is used in `render` instead.

Problematic:
```js
class Foo extends React.Component {
    render() {
        if (typeof window === 'undefined') {
            return null;
        }
        return <Bar />;
    }
}
```

Better:
```js
class Foo extends React.Component {
    constructor() {
        this.state = {showBar: false};
    }
    componentDidMount() {
        this.setState({showBar: true});
    }
    render() {
        if (!this.state.showBar) {
            return null;
        }
        return <Bar />;
    }
}
```

**Comment by lingyan** [comment]
:+1:

**Comment by mridgway** [comment]
Also good to check for in render: `Date.now()`, `new Date()`, `Math.random()`.

**Comment by lucasmotta** [comment]
Sorry to revive this issue, but I was thinking that it would be useful to have this rule for other Lifecycle events too. Like the `componentWillMount`, `componendDidUnmount`, etc...

**Comment by ljharb** [comment]
Absolutely, this should be separately configurable for every lifecycle method (defaulting to "on" for all) as well as separately configurable to allow an optional list of globals per lifecycle method.
